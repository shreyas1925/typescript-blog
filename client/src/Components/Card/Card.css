
.app--blocks-block {
    width: 320px;
    /* height: 580px; */
    margin: 10px;
    border: 1px solid #adb5bd;
    /* overflow-y: scroll; */
    height:fit-content

}
.app--blocks-flex{
    min-height: max-content;
    flex:1
}
.image--block{
    height: 50%;
    width:100%;
}

img{
    width: 100%;
    height: 16vw;
    object-fit: cover;
}

.block--timestamp{
    display: flex;
    justify-content: space-between;
    padding: 12px;
    font-size:small;
    font-weight: 300;
    color:#adb5bd
}

.block--description{
    padding: 12px;
    flex:1
}

.block--description > h2{
    margin-bottom: 10px;
    font-weight: 100;
    overflow: hidden;
   text-overflow: ellipsis;
   display: -webkit-box;
   -webkit-line-clamp: 2; 
           line-clamp: 2; 
   -webkit-box-orient: vertical;
}

.block--description > p{
    color:#6c757d;
    overflow: hidden;
   text-overflow: ellipsis;
   display: -webkit-box;
   -webkit-line-clamp: 3; 
           line-clamp: 3; 
   -webkit-box-orient: vertical;
}
.block--icons{
    /* margin-top: 10px; */
    display: flex;
    justify-content: space-between;
    padding: 12px;
    margin-bottom: -10;
}


.block--icons img{
    width: 20px;
    height: 20px;
}

.block--icons img:hover{
    cursor: pointer;
    width: 20px;
    height: 20px;
}


.block--clap{
    display: flex;
}

.block--clap p{
    margin-left: 3px;
    color:#6c757d
}

hr{
    margin: auto;
    width: 92% !important;
}


