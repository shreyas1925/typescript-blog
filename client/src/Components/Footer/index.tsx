import React from 'react'
import "./Footer.css"

const Footer = ():JSX.Element => {
  return (
    <footer data-testid='footer'>
        <h3>&amp;artifact.com 2019</h3>
    </footer>
  )
}

export default Footer