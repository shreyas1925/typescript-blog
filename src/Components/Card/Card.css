
.app--blocks-block {
    width: 320px;
    height: 480px;
    margin: 10px;
    border: 1px solid #adb5bd;
}

.image--block{
    height: 50%;
    width:100%;
}

img{
    width: 100%;
    height: 16vw;
    object-fit: cover;
}

.block--timestamp{
    display: flex;
    justify-content: space-between;
    padding: 12px;
    font-size:small;
    font-weight: 300;
    color:#adb5bd
}

.block--description{
    padding: 12px;
}

.block--description h2{
    margin-bottom: 10px;
    font-weight: 100;
}

.block--description p{
    color:#6c757d
}
.block--icons{
    margin-top: 10px;
    display: flex;
    justify-content: space-between;
    padding: 12px;
}


.block--icons img{
    width: 20px;
    height: 20px;
}

.block--icons img:hover{
    cursor: pointer;
    width: 20px;
    height: 20px;
}


.block--clap{
    display: flex;
}

.block--clap p{
    margin-left: 3px;
    color:#6c757d
}

hr{
    margin: auto;
    width: 92% !important;
}


