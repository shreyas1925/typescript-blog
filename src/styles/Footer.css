footer{
    background-color: black;
    color: white;
    padding:15px;
    padding-left: 135px;   
}