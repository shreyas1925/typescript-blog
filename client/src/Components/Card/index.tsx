import {useState } from 'react'
import clapping from "../../../src/assets/Icons/clapping.svg"
import heartRed from "../../../src/assets/Icons/heart-red.svg"
import heartBlack from "../../../src/assets/Icons/heart-black.svg"
import "./Card.css"
import { getFormattedDateFromUtcDate } from '../../utils/common'
import { BlogData } from '../../types'
import makeRequest from '../../utils/makeRequest'
import { UPDATE_BLOG_DATA } from '../../constants/apiEndPoints'

// type CardProps = {
//     date:string,
//     readingTime:string,
//     title:string,
//     description:string,
//     claps:number,
//     liked:boolean, 
//     image:string,
//     // setCount :(value: number | ((prevVar: number) => number)) => void;
//     // setCount:Dispatch<SetStateAction<boolean>>;  
// }


interface BlogCardProps  {
    blogData: BlogData
}

const Card: React.FC<BlogCardProps> = ({blogData}) => {

//   const handleClick = () => {
//     setCount(count+1)
//   }

//   const handleToggle = () => {
//     setHeart(!heart)
//   }

  const [clapCount, setClapCount] = useState(blogData.claps);
  const [isLiked, setIsLiked] = useState(blogData.liked);

  const handleClap = async () => {
    try {
      await makeRequest(UPDATE_BLOG_DATA(blogData.id), {
        data: { claps: clapCount + 1 },
      });
      setClapCount(clapCount + 1);
    } catch (e) {
     console.log(e);
     
    }
  };

  const handleLike = async () => {
    try {
      await makeRequest(UPDATE_BLOG_DATA(blogData.id), {
        data: { liked: !isLiked },
      });
      setIsLiked(!isLiked);
    } catch (e) {
        console.log(e);
    }
  };
  console.log(blogData)

  const likedImage = isLiked ? heartRed : heartBlack
  return (
    
        <div className="app--blocks-block" data-testid='card'>
            
                <div>
                    <img src={blogData.image} alt=""/>
                </div>
                <div className="block--timestamp">
                    <p>{getFormattedDateFromUtcDate(blogData.date)}</p>
                    <p>{blogData.reading_time}</p>       
                </div>
                <div className="block--description">
                    <h2>{blogData.title}</h2>
                    <p>{blogData.description}</p>
                </div>
                <hr/>
            
            <div className="block--icons">
                <div className="block--clap">
                    <img src={clapping} alt="" data-testid='count-img' onClick={handleClap}/>
                    <p data-testid='count'>{clapCount}</p>
                </div>
                <div>
                    <img src={likedImage} data-testid='img' onClick={() => handleLike()} alt=""/> 
                </div>
            </div>
        </div>
   
  )
}

export default Card