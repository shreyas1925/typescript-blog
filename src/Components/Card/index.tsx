import {useState } from 'react'
import clapping from "../../../src/Icons/clapping.svg"
import heartRed from "../../../src/Icons/heart-red.svg"
import heartBlack from "../../../src/Icons/heart-black.svg"
import "./Card.css"

type CardProps = {
    date:string,
    readingTime:string,
    title:string,
    description:string,
    claps:number,
    liked:boolean, 
    image:string,
    // setCount :(value: number | ((prevVar: number) => number)) => void;
    // setCount:Dispatch<SetStateAction<boolean>>;  
}
const Card = (props:CardProps):JSX.Element => {

  const [count,setCount] = useState(3)
  const [heart,setHeart] = useState(true)

  const handleClick = () => {
    setCount(count+1)
  }

  const handleToggle = () => {
    setHeart(!heart)
  }

  const likedImage = heart ? heartRed : heartBlack
  return (
    
        <div className="app--blocks-block">
            <div>
                <img src={props.image} alt=""/>
            </div>
            <div className="block--timestamp">
                <p>{props.date}</p>
                <p>{props.readingTime}</p>       
            </div>
            <div className="block--description">
                <h2>{props.title}</h2>
                <p>{props.description}</p>
            </div>
            <hr/>
            <div className="block--icons">
                <div className="block--clap">
                    <img src={clapping} alt="" onClick={handleClick}/>
                    <p>{count}</p>
                </div>
                <div>
                <img src={likedImage} onClick={() => handleToggle()} alt=""/> 
                </div>
            </div>
        </div>
   
  )
}

export default Card