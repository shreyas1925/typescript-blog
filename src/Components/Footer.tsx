import React from 'react'
import "../styles/Footer.css"

const Footer = ():JSX.Element => {
  return (
    <footer>
        <h3>&amp;artifact.com 2019</h3>
    </footer>
  )
}

export default Footer