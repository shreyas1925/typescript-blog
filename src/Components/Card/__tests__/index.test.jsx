import { fireEvent, render, screen } from "@testing-library/react"
import Card from ".."
import NavBar from "../../Navbar";
import Footer from "../../Footer";

describe('Card',()=>{
    
    it("should render correctly forming a snapshot", () => {
        const { asFragment } = render(<Card />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly without crashing',()=>{   
        render(<Card />)
        const cardText  = screen.getByTestId('card',{exact:false})
        expect(cardText).toBeTruthy()
    })

    describe('Heart', () => {
        it('should be black heart before clicking heart icon',()=>{   
            render(<Card/>);
            const ImageEvent = screen.getByTestId('img');
            expect(ImageEvent).toHaveAttribute('src', 'heart-black.svg');
        })
        it('should be red heart after clicking heart icon',()=>{   
            render(<Card/>);
            const ImageEvent = screen.getByTestId('img');
            fireEvent.click(ImageEvent);
            expect(ImageEvent).toHaveAttribute('src', 'heart-red.svg');
        })
    })

    describe('Claps Counter', () => {
        it('should be 0 before clicking clap icon',()=>{   
            render(<Card/>);
            const countEvent = screen.getByTestId('count');
            expect(countEvent).toHaveTextContent('0'); 
        })
        it('should be 1 when claps icon clicked once',()=>{   
            render(<Card/>);
            const countImageEvent = screen.getByTestId('count-img');
            const countEvent = screen.getByTestId('count');
            fireEvent.click(countImageEvent);
            expect(countEvent).toHaveTextContent('1');
        })
        it('should be 2 when claps icon clicked twice',()=>{   
            render(<Card/>);
            const countImageEvent = screen.getByTestId('count-img');
            const countEvent = screen.getByTestId('count');
            fireEvent.click(countImageEvent);
            fireEvent.click(countImageEvent);
            expect(countEvent).toHaveTextContent('2');
        })

        // it will check for 100,1000's also
        it('should be more than 1 when claps icon clicked more than once',()=>{   
            render(<Card/>);
            const countImageEvent = screen.getByTestId('count-img');
            const countEvent = screen.getByTestId('count');
            fireEvent.click(countImageEvent);
            fireEvent.click(countImageEvent);
            fireEvent.click(countImageEvent);
            expect(countEvent).toHaveTextContent('3');
        })
    })  
})

describe('Navbar',()=>{
    it('should render correctly without crashing',()=>{
        render(<NavBar/>)
        const navbarElement = screen.getByTestId('navbar')
        expect(navbarElement).toBeTruthy()
    })
    it('should render correctly without crashing forming a snapshot',()=>{
        const {asFragment} = render(<NavBar/>)
        expect(asFragment()).toMatchSnapshot()   
    })
})

describe('Footer',()=>{
    it('should render correctly without crashing',()=>{
        render(<Footer/>)
        const footerElement = screen.getByTestId('footer')
        expect(footerElement).toBeTruthy()
    })
    it('should render correctly without crashing forming a snapshot',()=>{
        const {asFragment} = render(<Footer/>)
        expect(asFragment()).toMatchSnapshot()   
    })
})