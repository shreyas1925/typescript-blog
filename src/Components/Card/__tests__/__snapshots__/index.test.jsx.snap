// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render correctly forming a snapshot 1`] = `
<DocumentFragment>
  <div
    class="app--blocks-block"
    data-testid="card"
  >
    <div>
      <img
        alt=""
      />
    </div>
    <div
      class="block--timestamp"
    >
      <p />
      <p />
    </div>
    <div
      class="block--description"
    >
      <h2 />
      <p />
    </div>
    <hr />
    <div
      class="block--icons"
    >
      <div
        class="block--clap"
      >
        <img
          alt=""
          data-testid="count-img"
          src="clapping.svg"
        />
        <p
          data-testid="count"
        >
          0
        </p>
      </div>
      <div>
        <img
          alt=""
          data-testid="img"
          src="heart-black.svg"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Footer should render correctly without crashing forming a snapshot 1`] = `
<DocumentFragment>
  <footer
    data-testid="footer"
  >
    <h3>
      &artifact.com 2019
    </h3>
  </footer>
</DocumentFragment>
`;

exports[`Navbar should render correctly without crashing forming a snapshot 1`] = `
<DocumentFragment>
  <header
    data-testid="navbar"
  >
    <div
      class="logo--header"
    >
      <h1>
        The Artifact
      </h1>
      <h3>
        Culture & Art Blog
      </h3>
    </div>
    <nav>
      <ul>
        <li>
          Blog
        </li>
        <li>
          About
        </li>
        <li>
          Contact
        </li>
      </ul>
    </nav>
  </header>
</DocumentFragment>
`;
